src/main/kotlin/com/asarkar/spring/test/redis/SocketUtils.kt

```diff
@@ -0,0 +1,153 @@
+     * @throws IllegalStateException if no available port could be found
+     */
+    /**
+     * Find an available TCP port randomly selected from the range
+     * [`minPort`, {@value #PORT_RANGE_MAX}].
+     * @param minPort the minimum port number
+     * @return an available TCP port number
+     * @throws IllegalStateException if no available port could be found
+     */
+    /**
+     * Find an available TCP port randomly selected from the range
+     * [{@value #PORT_RANGE_MIN}, {@value #PORT_RANGE_MAX}].
+     * @return an available TCP port number
+     * @throws IllegalStateException if no available port could be found
+     */
+    @JvmOverloads
+    fun findAvailableTcpPort(minPort: Int = PORT_RANGE_MIN, maxPort: Int = PORT_RANGE_MAX): Int {
+        return SocketType.TCP.findAvailablePort(minPort, maxPort)
+    }
+
+    private enum class SocketType {
+        TCP {
+            override fun isPortAvailable(port: Int): Boolean {
+                try {
+                    val serverSocket = ServerSocketFactory.getDefault().createServerSocket(
+                        port,
+                        1,
+                        InetAddress.getByName("localhost")
+                    )
+                    serverSocket.close()
+                    return true
+                } catch (ex: Exception) {
+                    return false
+                }
+            }
+        };
+
+        /**
+         * Determine if the specified port for this `SocketType` is
+         * currently available on `localhost`.
+         */
+        protected abstract fun isPortAvailable(port: Int): Boolean
+
+        /**
+         * Find a pseudo-random port number within the range
+         * [`minPort`, `maxPort`].
+         * @param minPort the minimum port number
+         * @param maxPort the maximum port number
+         * @return a random port number within the specified range
+         */
+        private fun findRandomPort(minPort: Int, maxPort: Int): Int {
+            val portRange = maxPort - minPort
+            return minPort + random.nextInt(portRange + 1)
+        }
+
+        /**
+         * Find an available port for this `SocketType`, randomly selected
+         * from the range [`minPort`, `maxPort`].
+         * @param minPort the minimum port number
+         * @param maxPort the maximum port number
+         * @return an available port number for this socket type
+         * @throws IllegalStateException if no available port could be found
+         */
+        fun findAvailablePort(minPort: Int, maxPort: Int): Int {
+            Assert.isTrue(minPort > 0, "'minPort' must be greater than 0")
+            Assert.isTrue(maxPort >= minPort, "'maxPort' must be greater than or equal to 'minPort'")
+            Assert.isTrue(maxPort <= PORT_RANGE_MAX, "'maxPort' must be less than or equal to $PORT_RANGE_MAX")
+
+            val portRange = maxPort - minPort
+            var candidatePort: Int
```

Comment:
This code is a conversion of the existing 
[org/springframework/util/SocketUtils.java](https://github.com/spring-projects/spring-framework/blob/5.3.x/spring-core/src/main/java/org/springframework/util/SocketUtils.java) code to Kotlin.
I have not been able to check whether it is thread safe in multithreading.

Reply:
@asarkar Could you please check the PR again?

Reply:
Thanks, but I would like to see some tests for SocketUtils. If Spring has some, feel free to use those, but otherwise, we need new tests.

Reply:
@asarkar The current PR was created because SocketUtils in Spring is not available.
Starting with Spring 6, SocketUtils.java used by asarkar has been removed from the code.

I will add the test code for SocketUtils mentioned above soon.